internal/scheduler/scheduler.go

```diff
@@ -0,0 +1,181 @@
+// Package scheduler implements a simple scheduler that triggers the next item when its due time is
+// reached based on the list of upcoming items.
+package scheduler
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/kopia/kopia/internal/clock"
+	"github.com/kopia/kopia/internal/ctxutil"
+	"github.com/kopia/kopia/repo/logging"
+)
+
+var log = logging.Module("scheduler")
+
+// GetItemsFunc is a callback that returns items for the scheduler to consider.
+type GetItemsFunc func(ctx context.Context, now time.Time) []Item
+
+// Item is a describes item that can be scheduled with a function invoked the next time the item is due.
+type Item struct {
+	Description string
+	NextTime    time.Time
+	Trigger     func()
+}
+
+// Scheduler is a scheduler triggers the next upcoming events based on the list of items returned by listItems().
+type Scheduler struct {
+	TimeNow func() time.Time
+	Debug   bool
+
+	refreshRequested chan string
+	getItems         GetItemsFunc
+	closed           chan struct{}
+	wg               sync.WaitGroup
+}
+
+// Options provides options for the scheduler.
+type Options struct {
+	TimeNow        func() time.Time
+	Debug          bool
+	RefreshChannel chan string
```

Comment:
Wondering, why is the channel provided to the scheduler? Can't it be owned by the scheduler and instead returned by a function? Or does it even need to be exposed at all? Can it be a function that is called on the scheduler such as:
`s.DoRefresh(string) error` or similar?

Reply:
I agree with @julio-lopez  comment here. A function would prevent the channel from being exposed to the caller. Similar to how Stop() works.

Reply:
the channel needs to be external, since it simplifies handling refreshes while we recreated - i was struggling with resolving a deadlock here and this was the only reliable choice I found, we can change it however we want since it's internal.